tests/components/life360/test_config_flow.py

```diff
@@ -0,0 +1,332 @@
+TEST_PW_2 = "password_2"
+TEST_PW_3 = "password_3"
+TEST_AUTHORIZATION = "authorization_string"
+TEST_AUTHORIZATION_2 = "authorization_string_2"
+TEST_AUTHORIZATION_3 = "authorization_string_3"
+TEST_MAX_GPS_ACCURACY = "300"
+TEST_DRIVING_SPEED = "18"
+TEST_SHOW_DRIVING = True
+
+USER_INPUT = {CONF_USERNAME: TEST_USER, CONF_PASSWORD: TEST_PW}
+
+TEST_CONFIG_DATA = {
+    CONF_USERNAME: TEST_USER,
+    CONF_PASSWORD: TEST_PW,
+    CONF_AUTHORIZATION: TEST_AUTHORIZATION,
+}
+TEST_CONFIG_DATA_2 = {
+    CONF_USERNAME: TEST_USER,
+    CONF_PASSWORD: TEST_PW,
+    CONF_AUTHORIZATION: TEST_AUTHORIZATION_2,
+}
+TEST_CONFIG_DATA_3 = {
+    CONF_USERNAME: TEST_USER,
+    CONF_PASSWORD: TEST_PW_3,
+    CONF_AUTHORIZATION: TEST_AUTHORIZATION_3,
+}
+TEST_DEF_OPTIONS = {
+    CONF_MAX_GPS_ACCURACY: None,
+    CONF_DRIVING_SPEED: None,
+    SHOW_DRIVING: False,
+}
+
+USER_OPTIONS = {
+    "limit_gps_acc": True,
+    CONF_MAX_GPS_ACCURACY: TEST_MAX_GPS_ACCURACY,
+    "set_drive_speed": True,
+    CONF_DRIVING_SPEED: TEST_DRIVING_SPEED,
+    SHOW_DRIVING: TEST_SHOW_DRIVING,
+}
+TEST_OPTIONS = {
+    CONF_MAX_GPS_ACCURACY: float(TEST_MAX_GPS_ACCURACY),
+    CONF_DRIVING_SPEED: float(TEST_DRIVING_SPEED),
+    SHOW_DRIVING: TEST_SHOW_DRIVING,
+}
+
+
+@pytest.fixture(name="life360", autouse=True)
+def life360_fixture():
+    """Mock life360."""
+    with patch("homeassistant.components.life360.config_flow.get_life360_api"), patch(
+        "homeassistant.components.life360.async_setup_entry", return_value=True
+    ):
+        yield
+
+
+async def _start_flow(hass):
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert not result["errors"]
+
+    return result
+
+
+def _authorization_login_error(hass, api, username, password, errors):
+    errors["base"] = "invalid_auth"
+    return None
+
+
+def _test_config_entry(hass, state=None):
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        data=TEST_CONFIG_DATA,
+        version=1,
+        state=state,
+        options=TEST_DEF_OPTIONS,
+        unique_id=TEST_USER.lower(),
+    )
+    config_entry.add_to_hass(hass)
+    return config_entry
+
+
+async def test_show_form(hass):
+    """Test that the form is served with no input."""
+    result = await _start_flow(hass)
+
+    schema = result["data_schema"].schema
+    assert set(schema.keys()) == set(USER_INPUT.keys())
+    # username and password fields should be empty.
+    keys = list(schema.keys())
+    for key in USER_INPUT:
+        assert keys[keys.index(key)].default == vol.UNDEFINED
+
+
+async def test_config_flow_success(hass):
+    """Test a successful config flow."""
+    result = await _start_flow(hass)
+    flow_id = result["flow_id"]
+
+    with patch(
+        "homeassistant.components.life360.config_flow.get_life360_authorization",
```

Comment:
I really want to add tests for more of the code, not just the config flow. In that case I'd be patching the api so as to avoid having to actually communicate with the real server. It would seem more appropriate to fully test out the response to the api in those tests, e.g., in a test_helpers.py file. But adding tests for more of the code seems like that should go into another PR once we're done with this one. And since I'd be checking the behavior of the code in helpers.py elsewhere, it seemed more appropriate, when testing config_flow.py, to patch the code it directly calls instead of some indirect code.

Reply:
The argument about helpers is moot as we should move that function to the config flow.

Reply:
True.